fluid/DeepQNetwork/expreplay.py

```diff
@@ -0,0 +1,51 @@
+#-*- coding: utf-8 -*-
+#File: expreplay.py
+
+from collections import namedtuple
+import numpy as np
+
+Experience = namedtuple('Experience', ['state', 'action', 'reward', 'isOver'])
+
+class ReplayMemory(object):
+    def __init__(self, max_size, state_shape):
+        self.max_size = int(max_size)
+        self.state_shape = state_shape
+
+        self.state = np.zeros((self.max_size, ) + state_shape, dtype='float32')
+        self.action = np.zeros((self.max_size, ), dtype='int32')
+        self.reward = np.zeros((self.max_size, ), dtype='float32')
+        self.isOver = np.zeros((self.max_size, ), dtype='bool')
+
+        self._curr_size = 0
+        self._curr_pos = 0
+
+    def append(self, exp):
+        if self._curr_size < self.max_size:
+            self._assign(self._curr_pos, exp)
+            self._curr_size += 1
+        else:
+            self._assign(self._curr_pos, exp)
+        self._curr_pos = (self._curr_pos + 1) % self.max_size
+
+    def _assign(self, pos, exp):
+        self.state[pos] = exp.state
+        self.action[pos] = exp.action
+        self.reward[pos] = exp.reward
+        self.isOver[pos] = exp.isOver
+
+    def __len__(self):
+        return self._curr_size
+
+    def sample(self, batch_idx):
+        for i, idx in enumerate(batch_idx):
+            while (idx + 1) % self._curr_size == self._curr_pos:
```

Comment:
I understand that you want to avoid the starting position of the circular buffer because the next state has not been written in the buffer yet. However, this while loop is confusing and inefficient. Perhaps when you initially sample batch_idx, you can sample from len(exp) - 1 and do an index mapping to avoid sampling the last exp in the buffer.

Reply:
This loop has two reasons:
1.the next state has not been written in the buffer yet
2.the next state has been written, but it's a state of another episode(long long ago...).